device/transport/mqtt/src/mqtt_base.ts

```diff
@@ -3,48 +3,128 @@
 
 'use strict';
 
-import * as querystring from 'querystring';
-import { Client as MqttClient } from 'mqtt';
-import { MqttReceiver } from './mqtt_receiver';
+import { EventEmitter } from 'events';
+import * as machina from 'machina';
+import { Client as MqttClient, IClientOptions, IClientPublishOptions, IClientSubscribeOptions } from 'mqtt';
 import * as dbg from 'debug';
 const debug = dbg('mqtt-common');
-import { errors, results, endpoint, Message } from 'azure-iot-common';
+import { errors, results, endpoint } from 'azure-iot-common';
 import { ClientConfig } from 'azure-iot-device';
 
-// tslint:disable-next-line:no-var-requires
-const packageJson = require('../package.json');
-
-/**
- * @private
- * @class           module:azure-iot-device-mqtt.MqttBase
- * @classdesc       Base MQTT transport implementation used by higher-level IoT Hub libraries.
- *                  You generally want to use these higher-level objects (such as [Mqtt]{@link module:azure-iot-device-mqtt.Mqtt})
- *                  rather than this one.
- *
- * @param {Object}  config      The configuration object derived from the connection string.
- */
 /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_004: [The `MqttBase` constructor shall instanciate the default MQTT.JS library if no argument is passed to it.]*/
 /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_005: [The `MqttBase` constructor shall use the object passed as argument instead of the default MQTT.JS library if it's not falsy.]*/
-export class MqttBase {
-  // should be private but isn't because it's used in mqtt.ts
-  client: MqttClient;
-
+export class MqttBase extends EventEmitter {
   private mqttprovider: any;
-  private _options: any;
-  private _receiver: any;
-  private _topicTelemetryPublish: string;
-  private _topicMessageSubscribe: string;
+  private _config: ClientConfig;
+  private _sdkVersionString: string;
+  private _mqttClient: MqttClient;
+  private _fsm: any;
 
-  constructor(mqttprovider?: any) {
+  constructor(sdkVersionString: string, mqttprovider?: any) {
+    super();
     this.mqttprovider = mqttprovider ? mqttprovider : require('mqtt');
+    this._sdkVersionString = sdkVersionString;
+
+    this._fsm = new machina.Fsm({
+      namespace: 'mqtt-base',
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err) => {
+            if (callback) {
+              callback(err);
+            } else {
+              if (err) {
+                this.emit('error', err);
+              }
+            }
+          },
+          connect: (callback) => this._fsm.transition('connecting', callback),
+          disconnect: (callback) => callback(),
+          /*Codes_SRS_NODE_COMMON_MQTT_BASE_16_020: [The `publish` method shall call the callback with a `NotConnectedError` if the connection hasn't been established prior to calling `publish`.]*/
+          publish: (topic, payload, options, callback) => callback(new errors.NotConnectedError()),
```

Comment:
>ew errors.NotConnectedError() [](start = 68, length = 29)

and this doesn't autoconnect because?

Reply:
because the connection information is passed in the `connect` call and we can't guarantee that `connect` has been called before `publish` is called when we're in the `disconnected` state.

---
In reply to: [140908660](https://github.com/Azure/azure-iot-sdk-node/pull/90#discussion_r140908660) [](ancestors = 140908660)